View Hierarchy (#33)
getsentry/rfcs: text/0033-view-hierarchy.md

* Start Date: 2022-11-02
* RFC Type: feature
* RFC PR: [#33](https://github.com/getsentry/rfcs/pull/33)
* RFC Status: -
* RFC Driver: [Dhiogo Ramos Brustolin](https://github.com/brustolin)

# Summary

Allow SDKs to send a textual representation of the application's UI view hierarchy. 

# Motivation

Being able to see what the user was looking at is a good way to find problems in an application. This RFC proposes a way to send information from the UI state of an application, known as "view hierarchy". Compared to screenshots the view hierarchy provides complementary information about the UI, e.g. relationships between UI widgets or additional attributes not visible on the screen.

Example View Hierarchies from PoC implementations:
* [Android](https://github.com/getsentry/sentry-java/pull/1998#issuecomment-1104806065)
* [iOS](https://github.com/getsentry/sentry-cocoa/pull/2044#issue-1329719543)
* [Flutter](https://github.com/ueman/sentry-dart-tools/blob/b850db587f0e099bed253d13055c88c03d536875/sentry_flutter_plus/lib/src/integrations/tree_walker_integration.dart)
* [Unity](https://github.com/getsentry/team-mobile/issues/64#issuecomment-1290868653)

# Proposal

> Capture the view hierarchy in our mobile SDKs and convert it to a common JSON representation. Add it as an [attachment](https://develop.sentry.dev/sdk/envelopes/#attachment) to the envelope. The `attachment_type` is set to `"event.viewhierarchy"` and the `content_type` is set to the `"application/json"` mime-type.

Reviewer: We need to add this type to the corresponding Relay type, otherwise the envelope will fail to parse afaict: https://github.com/getsentry/relay/blob/0586fa62eb3ee75697120ea6d89efe51c740f6ad/relay-server/src/envelope.rs#L362-L369.

Author: We just need to include the new type in the enum? Or there is more changes do to?

Reviewer: Yes that should be enough.

Reviewer: @jan-auer since we're not forward compatible in regards to attachment types, meaning that outdated external Relays will [fail to parse the envelope](https://github.com/getsentry/relay/blob/a9297f9cecb00b2caa09033a3c12438cc8c23b76/relay-server/src/envelope.rs#L1041) when there's a new attachment type, should we just use `AttachmentType::Attachment` here instead?

I noticed that for screenshots, there is [no special attachment type](https://github.com/getsentry/sentry/blob/1c9c4829a3d9f8957eb050b78539432792ae861f/src/sentry/models/eventattachment.py#L23-L30) either.

Author: Instead of attachment type we could maybe use a different mime type.

> Capture the view hierarchy in our mobile SDKs and convert it to a common JSON representation. Add it as an [attachment](https://develop.sentry.dev/sdk/envelopes/#attachment) to the envelope. The `attachment_type` is set to `"event.viewhierarchy"` and the `content_type` is set to the `"application/json"` mime-type.

Author: ```suggestion
Capture the view hierarchy in our front-end SDKs and convert it to a common JSON representation. Add it as an [attachment](https://develop.sentry.dev/sdk/envelopes/#attachment) to the envelope. The `attachment_type` is set to `"event.viewhierarchy"` and the `content_type` is set to the `"application/json"` mime-type.
```

# Options Considered
- Use a new envelope type instead of attachements: Whilst doable it's consideribly more work without adding any benefits. Using attachements has several advantages
    - Attachments are billable by default
    - We can immediately start sending and using this data, no need for large Relay updates
    - Less new code in the SDKs when compared to adding a new envelope item, so less hassle to maintain/test
    - We don't have to touch outcomes/rate limiting/processing code at all
    - We always can eventually migrate to a new envelope item type if we feel the need to, that option is still left open.
- The view hierarchy could be part of the Event, but because of the size limit, it is not recommended

# Drawbacks
- By utilizing attachments, [attachment scrubbing](https://docs.sentry.io/product/data-management-settings/scrubbing/attachment-scrubbing/) is relevant for any PII attributes in the view hierarchy
- Increases event payload size
- We've decided to not add the content of textfields, editables or labels because of PII.
    - We may evaluate an options to allow PII.

> ## Initial Proposal

Reviewer: Will the initial proposal remain a part of the final doc? It's a bit confusing to read.

Author: I think we can remove it.

This is how it should look like: 

```json
[
    {
        "type": "element_full_qualified_class_name", 
        "identifier": "element_identifier",
        "children": [...], //An array of ui elements.
        "width": 100, 
        "height": 100,
        "depth": 100, //if applies
        "x": 0,
        "y": 1,
        "z": 2, //if applies.
        "visible": true|false,
        "alpha": 1, //A float number from 0 to 1, where 0 means totally transparent, and 1 totally opaque.
        "{extra_properties}": "{property value}" //adicional information by platform
    }
]
```

This is just an example, each platform will define the required properties.

# Options Considered

- We could send the same information as attachment, but this impact on attachment quota. And also we need to rely on file name to determined whether an attachment is a view hierarchy or not, this is a problem we already have with [screenshots](https://develop.sentry.dev/sdk/features/#screenshots).
- The view hierarchy could be part of the Event, but because of the size limit, isn't recommended. 

# Drawbacks

- This can significantly increase the envelope size. 

> - Obfuscation is a problem for custom UI controls, we'll probably need ingestion to de-obfuscate in the server side.

Reviewer 2: Link this to https://docs.sentry.io/product/data-management-settings/scrubbing/attachment-scrubbing/

* RFC Driver: [Dhiogo Ramos Brustolin](https://github.com/brustolin)

# Summary

Allow SDKs to send a textual representation of the application's UI view hierarchy. 

# Motivation

Being able to see what the user was looking at is a good way to found out problems in the application. 
This RFC proposes a way to send information from the UI state of an application, known as "view hierarchy".

# Proposal 

The proposal is to create a new [envelope item](https://develop.sentry.dev/sdk/envelopes/) in the form of a json representation of the UI elements. 
This is how the envelope item header should look like:

```json
{"type":"view_hierarchy","length":...,"content_type":"application/json"}\n
```

And the envelope item payload will be a JSON array that contains a view hierarchy for each available screen/window. 
This is how it should look like: 

```json
[
    {
        "type": "element_class_name", 
        "identifier": "element_identifier",
        "children": [...], //An array of ui elements.
        "width": 100, 
        "height": 100,
        "depth": 100, //if 3D
        "x": 0,
        "y": 1,
        "z": 2, //if 3D.
        "visible": true|false,
        "alpha": 1,

>         "{extra_properties}": "{property value}" //adicional information by platform

Reviewer 2: What if we make a databag here, `Map<String, Any>`?

Author: It's possible. I'm not a fan. 
Property names can't duplicate, I don't see a reason we need an extra layer of properties.

Reviewer 2: No strong opinions, at least on Relay, we often have the pattern of this `data` databad that holds extra values.

Reviewer 3: I'd prefer being able to separate out dynamically added properties (i.e. between platforms) from common ones. I'm still trying to understand how this data will get surfaced in the product after ingestion, but knowing that certain keys are always there vs not will be helpful when working with the data type in the front end

Reviewer 4: I think we should always all for forwards compatibility for event schemas, so we allow for arbitrary keys, but we don't do anything with them.

* Start Date: 2022-11-02
* RFC Type: feature
* RFC PR: [#33](https://github.com/getsentry/rfcs/pull/33)
* RFC Status: -
* RFC Driver: [Dhiogo Ramos Brustolin](https://github.com/brustolin)

# Summary

Allow SDKs to send a textual representation of the application's UI view hierarchy. 

# Motivation

Being able to see what the user was looking at is a good way to found out problems in the application. 
This RFC proposes a way to send information from the UI state of an application, known as "view hierarchy".

# Proposal 

The proposal is to create a new [envelope item](https://develop.sentry.dev/sdk/envelopes/) in the form of a json representation of the UI elements. 
This is how the envelope item header should look like:

```json
{"type":"view_hierarchy","length":...,"content_type":"application/json"}\n
```

> And the envelope item payload will be a JSON array that contains a view hierarchy for each available screen/window. 

Reviewer 5: How does it differentiate between multiple windows and multiple root views?

Author: Each Window is a UI element in the array, and since properties are dynamic, the platform may choose different properties for Windows.

Mixed UI will have different properties, but still UI Elements in a hierarchy.

This is how it should look like: 

> ```json

Reviewer 5: Maybe use shorter keys to save some precious file size. Some UI hierarchies can become quite huge.

Author: Maybe `identifier` to `id`. but this properties will appear as information to the user. So we need to be clear.

[
    {
        "type": "element_class_name", 

>         "identifier": "element_identifier",

Author: Yes

        "children": [...], //An array of ui elements.
        "width": 100, 
        "height": 100,
        "depth": 100, //if 3D
        "x": 0,
        "y": 1,
        "z": 2, //if 3D.

>         "visible": true|false,

Author: This `visible` property is for `isHidden` in iOS, or `visibility` GONE or INVISIBLE for Android. 
Out of bound views may be determined by the position of the view relative to the size of the parent.

        "alpha": 1,

>         "{extra_properties}": "{property value}" //adicional information by platform

Author: Yes. Each platform can determine the required properties. The ones in this RFC are a small sample set.

* Start Date: 2022-11-02
* RFC Type: feature
* RFC PR: [#33](https://github.com/getsentry/rfcs/pull/33)
* RFC Status: -
* RFC Driver: [Dhiogo Ramos Brustolin](https://github.com/brustolin)

# Summary

Allow SDKs to send a textual representation of the application's UI view hierarchy. 

# Motivation

Being able to see what the user was looking at is a good way to found out problems in the application. 
This RFC proposes a way to send information from the UI state of an application, known as "view hierarchy".

# Proposal 

The proposal is to create a new [envelope item](https://develop.sentry.dev/sdk/envelopes/) in the form of a json representation of the UI elements. 
This is how the envelope item header should look like:

```json
{"type":"view_hierarchy","length":...,"content_type":"application/json"}\n
```

And the envelope item payload will be a JSON array that contains a view hierarchy for each available screen/window. 
This is how it should look like: 

```json

> [

Reviewer: Since this is a list, should each list element have information about what screen/window it refers to? Or is that covered by the `identifier`?

Author: This is an array of screen/windows. Every other ui element of each screen will be listed inside the `children` property.

    {

>         "type": "element_class_name", 

Reviewer 2: ```suggestion
        "type": "element_full_qualified_class_name", 
```

        "identifier": "element_identifier",
        "children": [...], //An array of ui elements.
        "width": 100, 
        "height": 100,
        "depth": 100, //if applies
        "x": 0,
        "y": 1,
        "z": 2, //if applies.
        "visible": true|false,
        "alpha": 1, //A float number from 0 to 1, where 0 means totally transparent, and 1 totally opaque.
        "{extra_properties}": "{property value}" //adicional information by platform
    }
]
```

This is just an example, each platform will define the required properties.

# Options Considered

- We could send the same information as attachment, but this impact on attachment quota. And also we need to rely on file name to determined whether an attachment is a view hierarchy or not, this is a problem we already have with [screenshots](https://develop.sentry.dev/sdk/features/#screenshots).
- The view hierarchy could be part of the Event, but because of the size limit, isn't recommended. 

# Drawbacks

- This can significantly increase the envelope size. 

> - Obfuscation is a problem for custom UI controls, we'll probably need ingestion to de-obfuscate in the server side.

Reviewer: Could you elaborate on this? what kind of obfuscation are we talking about?

Author: Java for example. A custom view may have its class name minified by pro-guard.

Reviewer: Right. In that case we might need to do a similar thing as we do for profiles: https://github.com/getsentry/sentry/blob/cf71af372677487d7d0a7fd8ac9dd092f9596cf4/src/sentry/profiles/task.py#L350-L360.

I guess the entries of the view hierarchy would then have to contain the fully qualified class name of the view (either as `identifier` or as additional field, see https://github.com/getsentry/rust-proguard/blob/4bc2cddd79cc8f7c9d14dda2f9f4c431e10c9424/src/mapper.rs#L169-L172.

Author: They will have, in the `type` property.

* Start Date: 2022-11-02
* RFC Type: feature
* RFC PR: [#33](https://github.com/getsentry/rfcs/pull/33)
* RFC Status: -
* RFC Driver: [Dhiogo Ramos Brustolin](https://github.com/brustolin)

# Summary

Allow SDKs to send a textual representation of the application's UI view hierarchy. 

# Motivation

Being able to see what the user was looking at is a good way to find problems in an application. This RFC proposes a way to send information from the UI state of an application, known as "view hierarchy". Compared to screenshots the view hierarchy provides complementary information about the UI, e.g. relationships between UI widgets or additional attributes not visible on the screen.

Example View Hierarchies from PoC implementations:
* [Android](https://github.com/getsentry/sentry-java/pull/1998#issuecomment-1104806065)
* [iOS](https://github.com/getsentry/sentry-cocoa/pull/2044#issue-1329719543)
* [Flutter](https://github.com/ueman/sentry-dart-tools/blob/b850db587f0e099bed253d13055c88c03d536875/sentry_flutter_plus/lib/src/integrations/tree_walker_integration.dart)
* [Unity](https://github.com/getsentry/team-mobile/issues/64#issuecomment-1290868653)

# Proposal

> Capture the view hierarchy in our front-end SDKs and convert it to a common JSON representation. Add it as an [attachment](https://develop.sentry.dev/sdk/envelopes/#attachment) to the envelope. The `attachment_type` is set to `"event.viewhierarchy"` and the `content_type` is set to the `"application/json"` mime-type.

Reviewer 4: Do we want to use `event.view_hierarchy` instead of `event.viewhierarchy`? I'll let you make the final call on that.

Author: ```suggestion
Capture the view hierarchy in our front-end SDKs and convert it to a common JSON representation. Add it as an [attachment](https://develop.sentry.dev/sdk/envelopes/#attachment) to the envelope. The `attachment_type` is set to `"event.view_hierarchy"` and the `content_type` is set to the `"application/json"` mime-type.
```

Reviewer: I agree that `event.view_hierarchy` is more readable, but we have a precedent with `event.applecrashreport`, so I would stick to that convention: https://github.com/getsentry/relay/blob/36ca3675eae2202478cc7e001c2d8c9f189581c8/relay-server/src/envelope.rs#L334-L336

Author: I see. Thanks for the reference

The attachment payload is a JSON structure containing the view hierarchy, here's an example:
```json
{
    "rendering_system": "compose", // or android_view_system, apple_uikit, apple_swiftui, unity, flutter, ...
    "windows": [
        {
            "type": "com.example.ui.FancyButton",
            "identifier": "R.id.button_logout",
            "children": [], // nested elements
            "width": 100.0,
            "height": 100.0,
            "depth": 100.0, // if applies
            "x": 0.0,
            "y": 0.0,
            "z": 2.0, // if applies.
            "visible": true|false,

>             "alpha": 1, // a float number from 0 to 1, where 0 means totally transparent, and 1 totally opaque.

Reviewer 4: ```suggestion
            "alpha": 1, // a float number from 0 to 1, where 0 means transparent, and 1 is opaque.
```

            "{extra_properties}": "{property value}" // additional platform-specific attributes
        }
    ]
}
```

Some remarks on the example:
 * `windows`: contains all visible windows, on mobile it's typically just one or two (e.g. if a dialog is open)
 * `type`: The fully qualified widget class name, this name may be obfuscated on certain platforms (e.g. Android release builds with proguard enabled)
 * `children` nests all child UI widgets, which then builds up the whole UI tree

> A typical Android/iOS view hierarchy for a single window consists of around 100-200 objects. Taking the attributes from the example above this generates a raw JSON file with a size of around 50KB. More complex view hierarchies for e.g. a Unity game may hold 1000-2000 objects, producing a JSON file of around 500KB.

Reviewer 4: ```suggestion
A typical Android/iOS view hierarchy for a single window consists of around 100-200 objects. Taking the attributes from the example above this generates a raw JSON file with a size of around 50KB. More complex view hierarchies, for example a Unity game, may hold 1000-2000 objects, producing a JSON file of around 500KB.
```

# Options Considered
- Use a new envelope type instead of attachments: Whilst doable it's considerably more work without adding any benefits. Using attachments has several advantages
    - Attachments are billable by default
    - We can immediately start sending and using this data, no need for large Relay updates
    - Less new code in the SDKs when compared to adding a new envelope item, so less hassle to maintain/test
    - We don't have to touch outcomes/rate limiting/processing code at all
    - We always can eventually migrate to a new envelope item type if we feel the need to, that option is still left open.
- The view hierarchy could be part of the Event, but because of the size limit, it is not recommended

# Drawbacks
- By utilizing attachments, [attachment scrubbing](https://docs.sentry.io/product/data-management-settings/scrubbing/attachment-scrubbing/) is relevant for any PII attributes in the view hierarchy

> - Increases event payload size

Author: Thanks. Proposal changed, didn't change the drawbacks.

> - Increases event payload size

Author: ```suggestion
```

Example View Hierarchies from PoC implementations:
* [Android](https://github.com/getsentry/sentry-java/pull/1998#issuecomment-1104806065)
* [iOS](https://github.com/getsentry/sentry-cocoa/pull/2044#issue-1329719543)
* [Flutter](https://github.com/ueman/sentry-dart-tools/blob/b850db587f0e099bed253d13055c88c03d536875/sentry_flutter_plus/lib/src/integrations/tree_walker_integration.dart)
* [Unity](https://github.com/getsentry/team-mobile/issues/64#issuecomment-1290868653)

# Proposal

Capture the view hierarchy in our mobile SDKs and convert it to a common JSON representation. Add it as an [attachment](https://develop.sentry.dev/sdk/envelopes/#attachment) to the envelope. The `attachment_type` is set to `"event.viewhierarchy"` and the `content_type` is set to the `"application/json"` mime-type.

The attachment payload is a JSON structure containing the view hierarchy, here's an example:
```json
{
    "rendering_system": "compose", // or android_view_system, apple_uikit, apple_swiftui, unity, flutter, ...

>     "screens": [

Author: @markushi I think windows is more correct, screens is more related to displays imo.

```suggestion
    "windows": [
```
